src/data/finsupp/basic.lean

```diff
@@ -776,6 +776,77 @@ end
 
 end graph
 
+/-! ### Declarations about `of_list_pairs` -/
+
+section of_list_pairs
+
+variable [has_zero M]
+
+/-- Builds a finitely supported function from the list of input and output pairs. Only the first
+pair with a given input is considered. This can be used to build a left inverse to `finsupp.graph`.
+-/
+noncomputable def of_list_pairs : list (α × M) → α →₀ M
```

Comment:
Does it make sense to define this as
```suggestion
noncomputable def alist.to_finsupp : alist α M → α →₀ M
```
Which essentially amounts to just proving that [`alist.lookup`](https://leanprover-community.github.io/mathlib_docs/data/list/alist.html#alist.lookup) has finite support.

Reply:
If supporting lists with duplicates is a feature, using `(list.lookup x).get_or_else 0` might still be a good idea as the function value, as then you don't need to prove anything about it besides the fact it's equal to `list.lookup`.

Reply:
I wasn't aware of `alist` - this is much closer to what I want. I'll redo this PR with it.

Reply:
Are you sure `list.lookup` isn't enough?

Reply:
I also wasn't aware of that, and it is really close to what I want, but I still want to implement it as a finitely supported function.

Reply:
That's fine; I'm suggesting that actually you might not need `alist`, not that there's no need for the finsupp version.

Having said that, I'm not sure this is really the right solution to your CNF motivation; can't you just use the recursor directly, or simply sum over the list?

Reply:
What I'm trying to do is to take the existing definition of `CNF` as a list of pairs, and retrieve a finitely supported function that returns the coefficient for a given exponent. I do still need the list definition though, since it's needed in order to state that the sum of terms in the Cantor Normal form equals the ordinal (we can't use `finset.sum` since ordinal addition isn't commutative).

Reply:
What's the use of a finitely supported function here? If addition isn't commutative, then presumably your function has an ordered support, which isn't something `finsupp` is able to represent.

Reply:
Which is why I want to define a finitely supported function alongside the list definition. The use is that there's theorems about the Cantor Normal form which are much nicer to state in terms of the function that returns the coefficient for an exponent. Here's two of them: `CNF_coeff b (o + o') ≤ CNF_coeff b o + CNF_coeff o'` and `CNF_coeff b (o ♯ o') = CNF_coeff b o ♯ CNF_coeff o'`. There's also a characterization of the coefficients of a normal multiplication, though it's much more difficult to state.

Reply:
> There's also a characterization of the coefficients of a natural multiplication, though it's much more difficult to state

I don't think it's difficult using finsupp, or is it? It's just multiplication in `add_monoid_algebra nat_ordinal` right? And add_monoid_algebra is defeq to finsupp. Aren't you aiming at constructing an N-alg_equiv between nat_ordinal and `add_monoid_algebra nat_ordinal`?

Reply:
I think you're right. Point stands, this is much easier to do if we have API for the coefficients of the CNF as a finsupp.